### PR TITLE
Add Support for Guild Scheduled Events

### LINKED ISSUE #165: Support for discord Scheduled Events
Please add support for discord scheduled events

### PR DESCRIPTION
**Pull request by PenguinNexus**
Please could this be reviewed and not blindly merged as not sure I've done what's needed here. When merged, this will resolve and close #165

### CODE REVIEW

File: src/Model/Guild/ScheduledEvent.php [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+     * the time the scheduled event will start
+     *
+     * @var \DateTimeImmutable|null
+     */
+    public $scheduled_start_time;
+
+    /**
+     * the time the scheduled event will end, required if entity_type is EXTERNAL
+     *
+     * @var \DateTimeImmutable|null
```
Comment (PenguinNexus) [inline src/Model/Guild/ScheduledEvent.php:62]: According to docs this should be a ISO8601 timestamp, is this the correct var to use here?

File: src/Model/Guild/ScheduledEvent.php [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+     * @var string
+     */
+    public $description;
+
+    /**
+     * the time the scheduled event will start
+     *
+     * @var \DateTimeImmutable|null
+     */
+    public $scheduled_start_time;
```
Comment (PenguinNexus) [inline src/Model/Guild/ScheduledEvent.php:57]: According to docs this should be a ISO8601 timestamp, is this the correct var to use here?

**Review by PenguinNexus — COMMENTED** [review]
Added comments to explain code - Assistance requested please

**Comment by rampmaster** [comment]
ping @aequasi

**Comment by cryptiklemur** [comment]
Can you please test the `dev-v9` branch to see if the guild scheduled events work there?

**Comment by baticag** [comment]
This would be great to see implemented

**Comment by LeigerGaming** [comment]
Would love to see this merged.

**Comment by cryptiklemur** [comment]
> Can you please test the dev-v9 branch to see if the guild scheduled events work there?

**Comment by LeigerGaming** [comment]
If I can work out how to specify that in composer.json, sure. Will take a look at it now.

**Comment by LeigerGaming** [comment]
Changed:
```
"restcord/restcord": "dev-develop",
```
Into this:
```
"restcord/restcord": "dev-v9",
```
And received the following error when attempting to update it:
```
Problem 1
    - Root composer.json requires restcord/restcord dev-v9, found restcord/restcord[dev-master, dev-develop, 0.1.0, ..., 0.5.0, v9.x-dev] but it does not match the constraint.
```

**Comment by LeigerGaming** [comment]
Managed to get it to update with `v9.x-dev` instead.

However, this sample code seems invalid:

```
$client->guild-scheduled-event->getGuildScheduledEvent($parameters);
```
... as seen in "Get Guild Scheduled Event.md"

How is it supposed to be called? As `guild-scheduled-event` does not work. And any more sensible values I attempt (camelCase or snake_case variants of the same thing) report an error, e.g. "No category with the name: guild_scheduled_event"

**Comment by LeigerGaming** [comment]
It looks like there's a similar issue with the audit log stuff, i.e. #152

**Comment by LeigerGaming** [comment]
And if I modify `service-description-v9.json` so that it works without the hyphens, I can't actually get it to successfully create an event via the API:

```php
$bot->guildscheduledevent->createGuildScheduledEvent([
    'guild.id' => $guildId,
    'name' => 'Testing',
]);
```
Error message:
```
There was an error executing the createGuildScheduledEvent command: Client error: `POST https://discord.com/api/v9/guilds/<snipped>/scheduled-events` resulted in a `400 Bad Request` response:
 (truncated...)
```
I've been unsuccessful in working out what the truncated part of the error message is.

**Comment by LeigerGaming** [comment]
Also, if I do provide some of the optional fields, I run into validation errors - which I suspect are due to Restcord incorrectly defining the types of these parameters?
```
Validation errors: [entity_metadata] must be of type entity metadata
[privacy_level] must be of type privacy level
[scheduled_start_time] must be of type ISO8601 timestamp
[scheduled_end_time] must be of type ISO8601 timestamp
[entity_type] must be of type entity type
```

I'm pretty confident that I'm defining at least a few of these correctly :\

Here's my options array:
```php
$bot->guildscheduledevent->createGuildScheduledEvent([
    'guild.id' => $guildId,
    'name' => 'Testing',
    'description' => 'This is a test description.',
    'scheduled_start_time' => Carbon::parse('8 January 2022 at 5pm')->toDateTimeImmutable(),
    'scheduled_end_time' => Carbon::parse('8 January 2022 at 7pm')->toDateTimeImmutable(),
    'privacy_level' => 2, // GUILD_ONLY
    'entity_type' => 3, // EXTERNAL
    'entity_metadata' => [
        'location' => 'https://example.com',
    ],
]);
```

**Comment by LeigerGaming** [comment]
I have found a solution for these issues and resolved it in #174

**Comment by SQKo** [comment]
Possibly Solved/Duplicate of #174

**Pull request closed without merging**